lib/Transforms/LLZKLoweringUtils.cpp

```diff
@@ -0,0 +1,179 @@
+                 << terminator->getName() << "'\n";
+    llvm_unreachable("compute() function must end with ReturnOp");
+  }
+  return retOp.getOperands().front();
+}
+
+Value rebuildExprInCompute(
+    Value val, FuncDefOp computeFunc, OpBuilder &builder, DenseMap<Value, Value> &memo
+) {
+  if (auto it = memo.find(val); it != memo.end()) {
+    return it->second;
+  }
+
+  if (auto barg = val.dyn_cast<BlockArgument>()) {
+    unsigned index = barg.getArgNumber();
+    Value mapped = computeFunc.getArgument(index - 1);
+    return memo[val] = mapped;
+  }
+
+  if (auto readOp = val.getDefiningOp<FieldReadOp>()) {
+    Value self = getSelfValueFromCompute(computeFunc);
+    Value rebuilt = builder.create<FieldReadOp>(
+        readOp.getLoc(), readOp.getType(), self, readOp.getFieldNameAttr().getAttr()
+    );
+    return memo[val] = rebuilt;
+  }
+
+  if (auto add = val.getDefiningOp<AddFeltOp>()) {
+    Value lhs = rebuildExprInCompute(add.getLhs(), computeFunc, builder, memo);
+    Value rhs = rebuildExprInCompute(add.getRhs(), computeFunc, builder, memo);
+    return memo[val] = builder.create<AddFeltOp>(add.getLoc(), add.getType(), lhs, rhs);
+  }
+
+  if (auto sub = val.getDefiningOp<SubFeltOp>()) {
+    Value lhs = rebuildExprInCompute(sub.getLhs(), computeFunc, builder, memo);
+    Value rhs = rebuildExprInCompute(sub.getRhs(), computeFunc, builder, memo);
+    return memo[val] = builder.create<SubFeltOp>(sub.getLoc(), sub.getType(), lhs, rhs);
+  }
+
+  if (auto mul = val.getDefiningOp<MulFeltOp>()) {
+    Value lhs = rebuildExprInCompute(mul.getLhs(), computeFunc, builder, memo);
+    Value rhs = rebuildExprInCompute(mul.getRhs(), computeFunc, builder, memo);
+    return memo[val] = builder.create<MulFeltOp>(mul.getLoc(), mul.getType(), lhs, rhs);
+  }
+
+  if (auto neg = val.getDefiningOp<NegFeltOp>()) {
+    Value operand = rebuildExprInCompute(neg.getOperand(), computeFunc, builder, memo);
+    return memo[val] = builder.create<NegFeltOp>(neg.getLoc(), neg.getType(), operand);
+  }
+
+  if (auto div = val.getDefiningOp<DivFeltOp>()) {
+    Value lhs = rebuildExprInCompute(div.getLhs(), computeFunc, builder, memo);
+    Value rhs = rebuildExprInCompute(div.getRhs(), computeFunc, builder, memo);
+    return memo[val] = builder.create<DivFeltOp>(div.getLoc(), div.getType(), lhs, rhs);
+  }
+
+  if (auto c = val.getDefiningOp<FeltConstantOp>()) {
+    return memo[val] = builder.create<FeltConstantOp>(c.getLoc(), c.getValue());
+  }
+
+  llvm::errs() << "Unhandled op in rebuildExprInCompute: " << val << '\n';
+  llvm_unreachable("Unsupported op kind");
+}
+
+LogicalResult checkForAuxFieldConflicts(StructDefOp structDef, StringRef prefix) {
+  bool conflictFound = false;
+
+  structDef.walk([&](FieldDefOp fieldDefOp) {
+    if (fieldDefOp.getName().starts_with(prefix)) {
+      (fieldDefOp.emitError() << "Field name '" << fieldDefOp.getName()
+                              << "' conflicts with reserved prefix '" << prefix << "'")
```

Comment:
nit: prefer char for size-1 string
```suggestion
                              << "' conflicts with reserved prefix '" << prefix << '\'')
```